File: rust/herb-linter/src/rules/herb_disable_comment_no_redundant_all.rs
use crate::herb_disable::parse_herb_disable_content;

use herb::nodes::ERBContentNode;
use herb::Visitor;

rule_visitor!(NoRedundantAllVisitor);
define_parser_rule!(
  HerbDisableCommentNoRedundantAllRule,
  "herb-disable-comment-no-redundant-all",
  Warning,
  NoRedundantAllVisitor
);

impl Visitor for NoRedundantAllVisitor {
  fn visit_erb_content_node(&mut self, node: &ERBContentNode) {
    let opening_tag = match &node.tag_opening {
      Some(token) => &token.value,
      None => return,
    };

    if opening_tag != "<%#" {
      return;
    }

    let content = match &node.content {
      Some(token) => &token.value,
      None => return,
    };

    let herb_disable = match parse_herb_disable_content(content) {
      Some(hd) => hd,
      None => return,
    };

    if !herb_disable.rule_names.contains(&"all".to_string()) {
      return;
    }

    if herb_disable.rule_names.len() <= 1 {
      return;
    }

    let all_detail = match herb_disable.rule_name_details.iter().find(|d| d.name == "all") {
      Some(d) => d,
      None => return,
    };

    let content_location = match &node.content {
      Some(token) => &token.location,
      None => return,
    };

    let location = all_detail.location(content_location).unwrap_or_else(|| node.location.clone());

    self.add_offense(
      "Using `all` with specific rules is redundant. Use `herb:disable all` by itself or list only specific rules.",
      location,
    );
  }
}

File: javascript/packages/linter/src/rules/herb-disable-comment-no-redundant-all.ts
import { ParserRule } from "../types.js"
import { HerbDisableCommentParsedVisitor } from "./herb-disable-comment-base.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ERBContentNode, ParseResult } from "@herb-tools/core"
import type { HerbDisableComment } from "../herb-disable-comment-utils.js"

class HerbDisableCommentNoRedundantAllVisitor extends HerbDisableCommentParsedVisitor {
  protected checkParsedHerbDisable(node: ERBContentNode, _content: string, herbDisable: HerbDisableComment): void {
    if (!herbDisable.ruleNames.includes("all")) return
    if (herbDisable.ruleNames.length <= 1) return

    const allDetail = herbDisable.ruleNameDetails.find(detail => detail.name === "all")
    if (!allDetail) return

    const location = this.createRuleNameLocation(node, allDetail)
    const message = `Using \`all\` with specific rules is redundant. Use \`herb:disable all\` by itself or list only specific rules.`

    this.addOffenseWithFallback(message, location, node)
  }
}

export class HerbDisableCommentNoRedundantAllRule extends ParserRule {
  static ruleName = "herb-disable-comment-no-redundant-all"
  static introducedIn = this.version("0.8.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new HerbDisableCommentNoRedundantAllVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}
